src/Workspaces/Core/Portable/Rename/Renamer.SyncNamespaceDocumentAction.cs

```diff
@@ -0,0 +1,118 @@
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Action that will sync the namespace of the document to match the folders property 
+        /// of that document
+        /// </summary>
+        internal class SyncNamespaceDocumentAction : RenameDocumentAction
+        {
+            private readonly AnalysisResult _analysis;
+
+            private SyncNamespaceDocumentAction(AnalysisResult analysis, OptionSet optionSet, ImmutableArray<ErrorResource> errors)
+                : base(errors, optionSet)
+            {
+                _analysis = analysis;
+            }
+
+            public override string GetDescription(CultureInfo? culture)
+             => WorkspacesResources.ResourceManager.GetString("Sync_namespace_to_folder_structure", culture ?? WorkspacesResources.Culture)!;
+
+            internal override async Task<Solution> GetModifiedSolutionAsync(Document document, CancellationToken cancellationToken)
+            {
+                var solution = document.Project.Solution;
+
+                if (!_analysis.SupportsSyncNamespace)
+                {
+                    return solution;
+                }
+
+                var changeNamespaceService = document.GetRequiredLanguageService<IChangeNamespaceService>();
+
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                var targetNamespace = BuildNamespaceFromFolders(document.Folders, syntaxFacts);
+
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                // Don't descend into anything other than top level declarations from the root.
+                // ChangeNamespaceService only controls top level declarations right now
+                var namespaceDeclarations = syntaxRoot.DescendantNodes(n => !syntaxFacts.IsDeclaration(n)).Where(n => syntaxFacts.IsNamespaceDeclaration(n));
+
+                foreach (var namespaceDeclaration in namespaceDeclarations)
+                {
+                    solution = await changeNamespaceService.ChangeNamespaceAsync(
+                        document,
+                        namespaceDeclaration,
+                        targetNamespace,
+                        cancellationToken).ConfigureAwait(false);
+
+                    document = solution.GetRequiredDocument(document.Id);
+                }
+
+                return solution;
+
+                static string BuildNamespaceFromFolders(
+                    IEnumerable<string> folders,
+                    ISyntaxFacts syntaxFacts)
+                {
+                    var parts = folders.SelectMany(folder => folder.Split(new[] { '.' }));
+
+                    // string.Empty is used to represent the global namespace
+                    // when calling ChangeNamespaceAsync
+                    return parts.All(syntaxFacts.IsValidIdentifier) ? string.Join(".", parts) : string.Empty;
```

Comment:
shouldn't we check this first?  if we get non-valid-identifiers, we should not offer this right?

Reply:
This is pulled directly from the sync namespace code action production, so it would be the same as the context menu refactoring we provide. We default to assuming global namespace if not all parts are valid identifiers

Reply:
> This is pulled directly from the sync namespace code action production, so it would be the same as the context menu refactoring we provide. We default to assuming global namespace if not all parts are valid identifiers

1. Why is that a desirable design though?   If my folders have stuff in them, why does that mean i'm syncing to root-namespace.  What if one folder is stuff like "src-assets/Foo/Bar/Baz".  Wouldn't it at least make sense for the namespace to be Foo.Bar.Baz, not empty.
2. if this is pulled from other code, we should have a shared helper.  if we update one, i woudl have no idea the other needed changing.

Reply:
> Why is that a desirable design though?

Honestly I'm not sure, I haven't gone through the deep analysis of all the tests in sync namespace to understand why. If we want to change that, fine, file a bug and we can discuss in a more appropriate forum and/or design review. 

> if this is pulled from other code

Sure, I can do that. I agree if this is the design propagating elsewhere would be hard to find without some shared code.

Reply:
Dug a little bit more into the design. I merged some of the logic, and if we can't determine a target namespace we won't provide an action at all. Thanks for pointing this out, it was a small divergence in behavior that wasn't intended.